dd-trace-core/src/main/java/datadog/trace/core/PendingTraceBuffer.java

```
@@ -0,0 +1,106 @@
+      try {
+        toClose.join();
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  public void flush() {
+    PendingTrace pendingTrace = queue.poll();
+    while (pendingTrace != null) {
+      pendingTrace.write();
+      pendingTrace = queue.poll();
+    }
+  }
+
+  private final class Worker implements Runnable {
+
+    @SneakyThrows
+    @Override
+    public void run() {
+      while (!Thread.interrupted()) {
+        PendingTrace pendingTrace = queue.poll();
+
+        if (pendingTrace == null) {
+          // Queue is empty.  Lets sleep and try again.
+          sleep();
+          continue;
+        }
+
+        if (pendingTrace.isEmpty()) {
+          // "write" it out to allow cleanup.
+          pendingTrace.write();
+          continue;
+        }
+
+        DDSpan rootSpan = pendingTrace.getRootSpan();
+
+        long finishTimestampMillis =
+            TimeUnit.NANOSECONDS.toMillis(rootSpan.getStartTime() + rootSpan.getDurationNano());
+        if (finishTimestampMillis + FORCE_SEND_DELAY_MS <= System.currentTimeMillis()) {
+          // Root span is getting old. We need to send the trace to avoid being discarded by agent.
+          pendingTrace.write();
+          continue;
+        }
+
+        long currentNanoTicks = Clock.currentNanoTicks();
+        long lastReferenced = pendingTrace.getLastReferenced();
+        long delta = currentNanoTicks - lastReferenced;
+
+        if (SEND_DELAY_NS <= delta) {
+          // Trace has been unmodified long enough, go ahead and write whatever is finished.
+          pendingTrace.write();
+        } else {
+          // Trace is too new.  Requeue it and sleep to avoid a hot loop.
+          queue.add(pendingTrace);
+          sleep();
+        }
+      }
+    }
+
+    private void sleep() {
+      try {
+        Thread.sleep(SLEEP_TIME_MS);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
```

I think we should take out this sleep (as mentioned elsewhere), but the handling of InterruptedException here is a bit odd.

We might as well let it bubble up and break the containing loop.  Right now, we catch it just to reset the interrupt status and then bail the loop later.

---

I did it this way to avoid printing a lot of thrown `InterruptedException`'s in the logs. Would you prefer the try on the outside of the loop?

---

Why would the InterruptedException show up in the logs?  You can catch inside the loop and break.

---

Without that catch, I was seeing the exception printed out quite a bit when I ran the tests.

---

I went ahead and moved the catch outside the loop and removed the `@SneakyThrows`, but it still needs to do something with the exception.  Would you prefer just ignore it since it doesn't matter if the interrupt flag is set at that point?

---

What else does it need to do something with the exception?  It interrupted the loop that's what it should do.  There's no requirement that we need to log or propagate.